README.md
<h3 align="center">PS-Inventory</h3>

<div align="center">

[![License](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)

</div>

## About
Fetch computer information and output to respective JSON file. Afterwards, build a reporting dashboard with PSWriteHTML.

## Getting Started
Since this is a small script, a simple `git clone` will get you a copy of the script up and running on your local machine for development and testing purposes.

This script relies on the
[ActiveDirectory](https://docs.microsoft.com/en-us/powershell/module/addsadministration/?view=win10-ps).
and
[PSWriteHTML](https://github.com/EvotecIT/PSWriteHTML)
modules.

```ps
Import-Module ActiveDirectory
Import-Module PSWriteHTML
```

It is also advantageous to install the RSAT tools for Active Directory.

```ps
Get-WindowsCapability -Name RSAT* -Online | Add-WindowsCapability -Online
```

## Usage

To start, run `.\PS-Inventory.ps1`. The text based menu will help you with the rest.

Once you have at least one computer scanned, run the `Build-Report.ps1` script. This will generate a report called `report.html` in the script root and will open automatically once finished. You can send this report to whoever via one single file!

If you want to modify the report builder, open the file and change a few configuration variables at the top (mainly for colors and ignoring some warnings).

Enjoy!

### Getting Help
The `Help` parameter will get you started on how the script behaves.

```ps
.\PS-Inventory.ps1 -Help
```

## Future Implementations

- [ ] Scan all computers in a specified OU
- [ ] Add report building into main script with parameter

## Resources

|   |
|---|
| [Parameters](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_advanced_parameters?view=powershell-6) |


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

This project adheres to simple versioning. Sort of (~~not really~~) a modified version of [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- No upcoming features since `1.0.0` release

## [1.0.0] - 2019-09-18

Build-Report.ps1
<# USER CONFIGURATION VARIABLES #>
$MinimumOSBuild = 1803 # Lowest build before conditional formatting is applied
$MaximumAge     = 5.9  # Maximum age before conditional formatting is applied
$UnsupportedOS  = 'Microsoft Windows 7 Professional' # Problem OS to conditionally format

# Domain applies to $DomainWarning, also possible to use (Get-ADDomain).Forest
$Domain         = 'YOURDOMAIN.org'

# If you don't want any warnings on the Devices tab, set values to $false
$MinimumOSBuildWarning  = $true  # Will warn if any devices are running builds lower than $MinimumOSBuild (above)
$UnsupportedOSWarning   = $true  # Will warn if any devices are running $UnsupportedOS (above)
$DomainWarning          = $false # Will warn if any devices are on a WORKGROUP

# Colors for Reports tab
$TotalDevicesColor = 'Green'
$AntivirusReportColor = 'Orange'
$WindowsEditionReportColor = 'Blue'
$WindowsBuildReportColor = 'Red'
############################################################################################################

# PSWriteHTML Module sources:
# https://evotec.xyz/easy-way-to-create-diagrams-using-powershell-and-pswritehtml/
# https://github.com/EvotecIT/PSWriteHTML

$Hosts   = (Get-ChildItem -Path "$PSScriptRoot\Hosts").FullName
$Content = Get-Content -Path $Hosts -Raw | ConvertFrom-Json

$Collection = [System.Collections.Generic.List[psobject]]::new()

foreach ($Item in $Content) {
    $Collection.Add([PSCustomObject]@{
        Hostname            = $Item.Hostname
        Device              = $Item.Device
        Type                = $Item.Type
        'Serial Number'     = $Item.Serial
        'Windows Edition'   = $Item.Edition
        'Windows Build'     = $Item.OS
        Memory              = $Item.Memory
        Domain              = $Item.Domain
        'Decimal Age'       = $Item.DecimalAge
        Age                 = $Item.Age
        'Reimaged Date'     = $Item.Reimaged
        'Excel Age Formula' = $Item.ExcelAge
        Antivirus           = $Item.AntiVirusProduct
    })
}

# Hastable enumertaion source : https://mjolinor.wordpress.com/2012/01/29/powershell-hash-tables-as-accumulators/

#region Counters
# These counters will break on legacy versions of PS-Inventory
# If you make core changes to what PS-Inventory outputs, then godspeed
$SerialCount            = [ordered]@{}
$DomainCount            = [ordered]@{}
$WindowsBuildCount      = [ordered]@{}
$AntivirusCount         = [ordered]@{}
$WindowsEditionCount    = [ordered]@{}
$DeviceTypeCount        = [ordered]@{}

$Collection | ForEach-Object {
    $SerialCount[$PSItem.'Serial Number']++
    $DomainCount[$PSItem.Domain]++
    $AntivirusCount[$PSItem.Antivirus]++
    $WindowsEditionCount[$PSItem.'Windows Edition']++
    $DeviceTypeCount[$PSItem.Type]++

    if ($PSItem.'Windows Build' -eq '6.3.9600') {
        $WindowsBuildCount['Windows 8.1 Pro']++
    }
    elseif ($PSItem.'Windows Build' -eq '6.1.7601') {
        $WindowsBuildCount['Windows 7 Professional']++
    }
    else {
        $WindowsBuildCount[$PSItem.'Windows Build']++  
    }
}

New-HTML -Name "Inventory Report" -FilePath "$PSScriptRoot\Report.html" -ShowHTML {
    New-HTMLTab -Name "Reports" {
        # Duplicate serial number warning
        foreach ($Serial in $SerialCount.GetEnumerator()) {
            if ($Serial.Value -eq 2) {
                New-HTMLPanel -Invisible {
                    New-HTMLToast -TextHeader "Duplicate Serial Numbers" -Text "Serial number [$($Serial.Name)] found twice in host files." -BarColorLeft OrangeRed -IconSolid info-circle -IconColor OrangeRed
                }
            }
        }

        # Start device reporting
        New-HTMLSection -Invisible {
            New-HTMLPanel -Invisible {
                New-HTMLChart -Title "Total Devices" {
                    New-ChartBarOptions -Vertical
                    New-ChartLegend -Names "Number of Devices" -Color $TotalDevicesColor
                    New-ChartBar -Name "Total Devices" -Value $Collection.Count
                }
            }

            New-HTMLPanel -Invisible {
                New-HTMLChart -Title "Device Types" {
                    New-ChartBarOptions -Vertical
                    New-ChartLegend -Names "Device Types" 
                    foreach ($Type in $DeviceTypeCount.GetEnumerator()) {
                        New-ChartPie -Name $Type.Name -Value $Type.Value
                    }
                }
            }
        } 

        New-HTMLChart -Title "Antivirus Report" {
            New-ChartLegend -Names "Antivirus Product Report" -Color $AntivirusReportColor
            foreach ($Antivirus in $AntivirusCount.GetEnumerator()) {
                New-ChartBar -Name $Antivirus.Name -Value $Antivirus.Value
            }
        }
        New-HTMLChart -Title "Windows Edition Report" {
            New-ChartBarOptions -Vertical
            New-ChartLegend -Names "Windows Edition" -Color $WindowsEditionReportColor
            foreach ($Edition in $WindowsEditionCount.GetEnumerator()) {
                New-ChartBar -Name $Edition.Name -Value $Edition.Value
            }
        }
        New-HTMLSection -Invisible {
            New-HTMLPanel -Invisible {
                New-HTMLChart -Title "Windows Build Report" {
                    New-ChartBarOptions -Vertical
                    New-ChartLegend -Names "Windows Build" -Color $WindowsBuildReportColor
                    foreach ($Build in $WindowsBuildCount.GetEnumerator()) {
                        New-ChartBar -Name $Build.Name -Value $Build.Value
                    }
                }
            }
        } 
    }
    
    New-HTMLTab -Name "Devices" {
        if ($UnsupportedOSWarning) {
            if ($WindowsBuildCount.'Windows 7 Professional' -gt 0) {
                New-HTMLToast -TextHeader "Windows 7 Support Ending Soon!" -Text "You have devices still running Windows 7.<br>
                <a href='https://support.microsoft.com/en-us/help/4057281/windows-7-support-will-end-on-january-14-2020 target='_blank''>Windows 7 EOL</a>" -BarColorLeft OrangeRed -IconSolid info-circle -IconColor OrangeRed
            }
        }
        
        if ($MinimumOSBuildWarning) {
            if ($WindowsBuildCount.GetEnumerator() | Where-Object {$_.Name -lt $MinimumOSBuild} ) {
                New-HTMLToast -TextHeader "Out of date devices!" -Text "You have devices still running builds lower than $MinimumOSBuild." -BarColorLeft OrangeRed -IconSolid info-circle -IconColor OrangeRed
            }
        }
        
        if ($DomainWarning) {
            if ($DomainCount.GetEnumerator() | Where-Object {$_.Name -eq 'WORKGROUP'}) {
                New-HTMLToast -TextHeader "Devices not joined to $Domain" -Text "You have devices still joined to a WORKGROUP." -BarColorLeft OrangeRed -IconSolid info-circle -IconColor OrangeRed
            }
        }
        
        New-HTMLSection -Name "All Device Information" {
            New-HTMLTable -HideFooter -DataTable $Collection {
                New-TableCondition -Name 'Decimal Age' -ComparisonType number -Operator gt -Value $MaximumAge -Color White -BackgroundColor Red -Row
                New-TableCondition -Name 'Windows Build' -ComparisonType number -Operator lt -Value $MinimumOSBuild -Color White -BackgroundColor Orange
                New-TableCondition -Name 'Windows Edition' -ComparisonType string -Operator eq -Value $UnsupportedOS -Color White -BackgroundColor Yellow
            }
        }
    }

    New-HTMLTab -Name "Reimaged Information" {
        New-HTMLCalendar {
            foreach ($Item in $Collection) {
                New-CalendarEvent -StartDate $Item.'Reimaged Date' -Title "$($Item.Hostname) reimaged"
            }
        }
    }
}

Get-Inventory.ps1
<#
.SYNOPSIS
Fetch computer information.

.DESCRIPTION
Fetch computer information and output to respective JSON file.

.PARAMETER Help
Prints help information.

.EXAMPLE
./Inventory.ps1 -Help

.NOTES
===========================================================================
    Created on:   	9/18/2019
    Updated on:     ***
	Author:    	    Matthew Fabrizio
	Organization: 	*** 
	Filename:     	Get-Inventory.ps1
===========================================================================
#>

# Prints help information.
param(
    [Parameter(Mandatory=$false,
           Position=0,
           HelpMessage="Prints help information.")]
    [switch]
    $Help
)

function Get-Help() {
    "
    AD Query Scan:
        The AD Query menu option allows a user to scan any number of devices starting with a certain string of text.
        For instance, say you want to scan all computers starting with a specific name of 'LAB-' and there are 20 computers ranging from LAB-01 to LAB-20
        The scan will ping all computers starting with that name and return the results, as well as, output all info to respective JSON files.

    Single and Loop Scan are pretty self explanatory; only requiring a hostname to be entered.

    Error Logging:
        If the script runs into an error, the information will be logged to a directory called Log in $PSScriptRoot
        This will contain information specific to what line of code actually failed.

    Excel Formatting:
        This script automatically copies all scanned devices to the clipboard in tab delimited format.
        This allows for simple pasting into Excel.

    Device Type:
        The device type is generated based on device name. The script is catered towards Dell devices and if anything isn't an Optiplex, it's deemed a laptop.
        It's not hard to change and other desktop models can be added if you really want, but overall it's sort of useless.
    "
}

function Write-LogEntry {
    param(
        [parameter(Mandatory=$true, HelpMessage="Value of log entry to be added.")]
        [ValidateNotNullOrEmpty()]
        [string] $Value,

        [parameter(Mandatory=$false, HelpMessage="Name of the log file.")]
        [ValidateNotNullOrEmpty()]
        [string] $FileName = "error.log",

        [Parameter(Mandatory=$false, HelpMessage="Prevents logging to STDOUT")]
        [switch] $OutNull
    )
    
    <# If no Log directory exists, create it #>
    if (!(Test-Path "$PSScriptRoot\Log" -PathType Container)) {
        New-Item -Path "$PSScriptRoot" -Name "Log" -ItemType Directory | Out-Null
    }

    <# Determine log file location #>
    $LogFilePath = Join-Path -Path "$PSScriptRoot" -ChildPath "Log\$($FileName)"

    <# Add value to log file #>
    try {
        if (!$OutNull) { Write-Host $Value }
        Out-File -InputObject "[$(Get-Date -Format g)] => $Value" -Append -NoClobber -Encoding Default -FilePath $LogFilePath -ErrorAction Stop
    }
    catch [System.Exception] {
        Write-Warning -Message "Unable to append log entry to $FileName file"
    }
}

function Get-Menu {
    Clear-Host

    Write-Host "`n----------- MENU -----------"
    Write-Host "[A] : AD Query Scan"
    Write-Host "[L] : Loop Scan"
    Write-Host "[S] : Single Scan"
    Write-Host "[Q] : Quit"
    Write-Host "----------------------------`n"
}

function Get-DeviceInfo() {
    param (
        # Computer(s) to be scanned
        [Parameter(Mandatory=$true)]
        [string[]]
        $ComputerName
    )
    
    # Demolish anything in the clipboard
    $Null | Set-Clipboard

    $Online = [System.Collections.ArrayList]::New()
    $DeviceArray = [System.Collections.ArrayList]::New()

    foreach ($Computer in $ComputerName.ToUpper()) {
        Write-Progress "Pinging $Computer"
        if (Test-Connection -ComputerName $Computer -Count 1 -Quiet) {
            Write-Host "Connection to [$Computer] successful." -ForegroundColor Green
            [Void]$Online.Add($Computer)
        }
        else { Write-LogEntry "$Computer unavailable." }
    }

    foreach ($Computer in $Online) {
        try {
            # Load relevant WMI classes
            $Win32_OperatingSystem = (Get-WmiObject -ComputerName $Computer -Class Win32_OperatingSystem -ErrorAction Stop)
            $Win32_ComputerSystem  = (Get-WmiObject -ComputerName $Computer -Class Win32_ComputerSystem -ErrorAction Stop)
            $Win32_Bios            = (Get-WmiObject -ComputerName $Computer -Class Win32_Bios -ErrorAction Stop)
            $Win32_PhysicalMemory  = (Get-WmiObject -ComputerName $Computer -Class Win32_PhysicalMemory -ErrorAction Stop)
            # $Win32_NetworkAdapterConfiguration = (Get-WmiObject -ComputerName $Computer -Class Win32_NetworkAdapterConfiguration -ErrorAction Stop)
            $Win32_NetworkAdapter = (Get-WmiObject -ComputerName $Computer -Class Win32_NetworkAdapter | Where-Object { $_.Description -notmatch 'wan miniport|microsoft isatap adapter|bluetooth|juniper|ras async adapter|virtual|apple|miniport|tunnel|debug|advanced-n|wireless-n|ndis' } -ErrorAction Stop)
            
            # Manufacturer / Physical
            $Manufacturer = $Win32_ComputerSystem.Manufacturer
            $Model        = $Win32_ComputerSystem.Model
            $Serial       = $Win32_Bios.SerialNumber
            $Memory       = $Win32_PhysicalMemory | Measure-Object -Property Capacity -Sum | ForEach-Object { "{0:N2}" -f ([math]::round(($_.Sum / 1GB),2)) }
            
            # NetBIOS
            $Hostname     = $Win32_OperatingSystem.CSName

            # Operating System
            $Edition      = $Win32_OperatingSystem.Caption
            $OS           = $Win32_OperatingSystem.Version
            
            # Lifespan
            $Age          = $Win32_Bios
            $ReimageDate  = $Win32_OperatingSystem

            # Networking
            $Domain       = $Win32_ComputerSystem.Domain
            $EthMAC       = ($Win32_NetworkAdapter | Where-Object {$_.NetConnectionID -like '*Ethernet*'}).MACAddress
            $WlpMAC       = ($Win32_NetworkAdapter | Where-Object {$_.NetConnectionID -like '*Wi-Fi*'}).MACAddress

            <# AD Queries #>
            # $Hostname = Get-ADComputer $Computer -Properties CN | Select-Object -ExpandProperty CN
            # $OS = Get-ADComputer $Computer -Properties OperatingSystem | Select-Object -ExpandProperty OperatingSystem
            # $IP = Get-ADComputer $Computer -Properties IPv4Address | Select-Object -ExpandProperty IPv4Address

            switch($OS){
                '10.0.10240' {$OS="1507"}
                '10.0.10586' {$OS="1511"}
                '10.0.14393' {$OS="1607"}
                '10.0.15063' {$OS="1703"}
                '10.0.16299' {$OS="1709"}
                '10.0.17134' {$OS="1803"}
                '10.0.17763' {$OS="1809"}
                '10.0.18362' {$OS="1903"}
            }

            # Mainly for Dell devices to remove Inc.
            $Manufacturer = [string]$Manufacturer.replace("Inc.","")
            $Make = $Manufacturer + $Model

            # Mainly for Dell devices, can add more desktop options if necessary
            if ($Model -like '*Optiplex*') { $Type = "Desktop" } 
            else { $Type = "Laptop" }

            # Calculate age and reimage date
            $Age = (New-TimeSpan -Start ($Age.ConvertToDateTime($Age.ReleaseDate).ToShortDateString()) -End $(Get-Date)).Days / 365
            $ReimageDate = ($ReimageDate.ConvertToDateTime($ReimageDate.InstallDate).ToString("MM/dd/yyyy"))

            <# Calculate Decimal Age to Date #>
            $CalculatedAge = ($Age * 365.25)
            $CalculatedAge = (Get-Date).AddDays(-$CalculatedAge).ToString("MM/dd/yyyy")

            # Calculate Excel formula based off TODAYs date
            $ExcelAge = "=ROUND(YEARFRAC(`"$CalculatedAge`", TODAY()), 2)"

            # Big ol object dump
            $Properties = [PSCustomObject]@{
                Hostname    = $Hostname
                Device      = $Make
                Type        = $Type
                Serial      = $Serial
                Edition     = $Edition
                OS          = $OS
                Memory      = $Memory
                Domain      = $Domain
                DecimalAge  = $Age
                Age         = $CalculatedAge
                Reimaged    = $ReimageDate
                ExcelAge    = $ExcelAge
            }

            # Add MAC if exists to object
            if ($EthMAC) { $Properties | Add-Member -NotePropertyName EthMAC -NotePropertyValue $EthMAC }
            if ($WlpMAC) { $Properties | Add-Member -NotePropertyName WlpMAC -NotePropertyValue $WlpMAC }

            # Store computer info in DeviceArray, append additional devices
            [Void]$DeviceArray.Add($Properties)

            # Output computer data to JSON in Hosts directory
            $Properties | ConvertTo-Json | Out-File $PSScriptRoot\Hosts\$Computer.json

            <# Prep specific properties for Excel pasting; tab delimited #>
            ($Properties | Select-Object * | ForEach-Object {
                $_.Device
                $_.Type
                $_.Hostname
                $_.Serial
                $_.Edition
                $_.OS
                $_.ExcelAge
                $_.Reimaged
            }) -join "`t" | Set-Clipboard -Append
        }
        catch [System.Exception] {
            $ExceptionLineNumber  = $PSItem.InvocationInfo.ScriptLineNumber
            $ExceptionLineContent = (Get-Content (Split-Path $MyInvocation.ScriptName -Leaf) -TotalCount $ExceptionLineNumber)[-1]
            $ExceptionMessage     = $PSItem.Exception.Message

            Write-LogEntry "[ERROR] : Device [$Computer]
            $ExceptionMessage
            Exception caugh on line $ExceptionLineNumber
            $ExceptionLineContent"
        }
    }
    
    <# Spicy STDOUT #>
    $DeviceArray
}

# If help, help please
if ($Help) { Get-Help; exit }

<# Remove Log directory if it exists; generate on error #>
if (Test-Path -Path "$PSScriptRoot\Log") {
    Remove-Item -Path "$PSScriptRoot\Log" -Recurse
}

<# If there is no Hosts directory to store JSON, create it #>
if (!(Test-Path -Path $PSScriptRoot\Hosts)) {
    New-Item -Path $PSScriptRoot -Name "Hosts" -ItemType Container > $null
}

do {
    <# Print the menu; Select the choice #>
    Get-Menu
    $Choice = Read-Host "Make a selection"

    switch ($Choice) {
        <# Prompt for AD search terms - only starting characters #>
        'a' { 
            $ADFilter = Read-Host "What computer would you like to search for?"
            $ADQuery = (Get-ADComputer -Filter "Name -like '$ADFilter*'" | Select-Object -ExpandProperty Name) -join ","
            $ADQuery = $ADQuery.Split(",").Trim(" ")
            Get-DeviceInfo -ComputerName $ADQuery
        }
        <# [l|L] Prompt for a computer to scan; exit on SIGINT #>
        <# [s|S] Prompt for a computer to scan; exit once complete #>
        { @('l','s') -contains $PSItem } {
            $hostnameExists = $false
            while (!$hostnameExists) {
                Write-Host "`nType 'stop|Stop|Ctrl+c' to exit.`n" -ForegroundColor Yellow
                $Computers = Read-Host "What computer would you like to scan?"
                if ($Computers -contains 'stop') { exit }
                $Computers = $Computers.Split(",").Trim(" ")
                Get-DeviceInfo -ComputerName $Computers
                if ($PSItem -eq 's') { $hostnameExists=$true }
            }
        }
        <# Log user quit prompt; Clear the screen; Exit the application #>
        'q' { Clear-Host; exit }
        default { Write-Host "Invalid menu choice" -ForegroundColor Red }  
    }
} while ($Choice -eq 'q')

PS-Inventory.ps1
<#
.SYNOPSIS
Fetch computer information.

.DESCRIPTION
Fetch computer information and output to respective JSON file.

.PARAMETER Help
Prints help information.

.EXAMPLE
./PS-Inventory.ps1 -Help

.NOTES
===========================================================================
    Created on:   	2019-09-18
	Author:    	    Matthew Fabrizio
	Organization: 	*** 
    Filename:     	PS-Inventory.ps1
===========================================================================

.LINK
https://github.com/matthewfabrizio
#>

# Requires -Module ActiveDirectory

# Prints help information.
[CmdletBinding()] param (
    [Parameter()]
    [switch]
    $Help,
    # Hosts storage
    [Alias('Path')]
    [ValidateScript({if ($PSItem){ Test-Path $PSItem}})]
    [string]
    $FilePath = "$PSScriptRoot"
)

# UNC paths are meh right now

function Get-Help() {
    Clear-Host

    "
    Welcome to PS-Inventory!

    AD Query Scan:
        The AD Query menu option allows a user to scan any number of devices starting with a certain string of text.
        For instance, say you want to scan all computers starting with a specific name of 'LAB-' and there are 20 computers ranging from LAB-01 to LAB-20
        The scan will ping all computers starting with that name and return the results, as well as, output all info to respective JSON files.

    Single and Loop Scan are pretty self explanatory; only requiring a hostname to be entered.

    Error Logging:
        If the script runs into an error, the information will be logged to a directory called Log in $PSScriptRoot
        This will contain information specific to what line of code actually failed.

    Excel Formatting:
        This script automatically copies all scanned devices to the clipboard in tab delimited format.
        This allows for simple pasting into Excel.

    Device Type:
        The device type is generated based on device name. The script is catered towards Dell devices and if anything isn't an Optiplex, it's deemed a laptop.
        It's not hard to change and other desktop models can be added if you really want, but overall it's sort of useless.
    "
}

function Show-Menu {
    # Clear-Host

    "`n----------- MENU -----------"
    "[A] : AD Query Scan"
    "[L] : Loop Scan"
    "[S] : Single Scan"
    "[Q] : Quit"
    "----------------------------`n"
}

function Get-DeviceInfo() {
    param (
        # Computer(s) to be scanned
        [Parameter(Mandatory)]
        [string[]]
        $ComputerName
    )
    
    # Demolish anything in the clipboard
    $Null | Set-Clipboard

    $Online = [System.Collections.ArrayList]::New()
    $DeviceArray = [System.Collections.ArrayList]::New()

    foreach ($Computer in $ComputerName.ToUpper()) {
        Write-Progress "Pinging $Computer"

        if (Test-Connection -ComputerName $Computer -Count 1 -Quiet) {
            Write-Host "Connection to [$Computer] successful." -ForegroundColor Green
            [Void]$Online.Add($Computer)
        }
        else { Write-Warning -Message "$Computer unavailable." }
    }

    foreach ($Computer in $Online) {
        try {
            # Load relevant WMI classes
            $Win32_OperatingSystem = (Get-WmiObject -ComputerName $Computer -Class Win32_OperatingSystem -ErrorAction Stop)
            $Win32_ComputerSystem  = (Get-WmiObject -ComputerName $Computer -Class Win32_ComputerSystem -ErrorAction Stop)
            $Win32_Bios            = (Get-WmiObject -ComputerName $Computer -Class Win32_Bios -ErrorAction Stop)
            $Win32_PhysicalMemory  = (Get-WmiObject -ComputerName $Computer -Class Win32_PhysicalMemory -ErrorAction Stop)
            $Win32_NetworkAdapter  = (Get-WmiObject -ComputerName $Computer -Class Win32_NetworkAdapter | Where-Object { $_.Description -notmatch 'wan miniport|microsoft isatap adapter|bluetooth|juniper|ras async adapter|virtual|apple|miniport|tunnel|debug|advanced-n|wireless-n|ndis' } -ErrorAction Stop)
            $AV = (Get-WmiObject -ComputerName $Computer -Namespace "root\SecurityCenter2" -Query "SELECT * FROM AntiVirusProduct").displayName
            $Win32_SystemEnclosure = (Get-WmiObject -ClassName Win32_SystemEnclosure -Namespace 'root\CIMV2' -Property ChassisTypes).ChassisTypes

            # Manufacturer / Physical
            $Manufacturer = $Win32_ComputerSystem.Manufacturer
            $Model        = $Win32_ComputerSystem.Model
            $Serial       = $Win32_Bios.SerialNumber
            $Memory       = $Win32_PhysicalMemory | Measure-Object -Property Capacity -Sum | ForEach-Object { "{0:N2}" -f ([math]::round(($_.Sum / 1GB),2)) }
            
            # NetBIOS
            $Hostname     = $Win32_OperatingSystem.CSName

            # Operating System
            $Edition      = $Win32_OperatingSystem.Caption
            $OS           = $Win32_OperatingSystem.Version
            
            # Lifespan
            $Age          = $Win32_Bios
            $ReimageDate  = $Win32_OperatingSystem

            # Networking
            $Domain       = $Win32_ComputerSystem.Domain
            $EthMAC       = ($Win32_NetworkAdapter | Where-Object {$_.NetConnectionID -like '*Ethernet*'}).MACAddress
            $WlpMAC       = ($Win32_NetworkAdapter | Where-Object {$_.NetConnectionID -like '*Wi-Fi*'}).MACAddress

            switch($OS){
                '10.0.10240' {$OS="1507"}
                '10.0.10586' {$OS="1511"}
                '10.0.14393' {$OS="1607"}
                '10.0.15063' {$OS="1703"}
                '10.0.16299' {$OS="1709"}
                '10.0.17134' {$OS="1803"}
                '10.0.17763' {$OS="1809"}
                '10.0.18362' {$OS="1903"}
            }

            # Mainly for Dell devices to remove Inc.
            $Manufacturer = [string]$Manufacturer.replace("Inc.","")
            $Make = $Manufacturer + $Model

            # Chassis type source values : https://docs.microsoft.com/en-us/windows/win32/cimwin32prov/win32-systemenclosure
            switch ($Win32_SystemEnclosure) {
                { @(3..7)  -contains $PSItem } { $Type = "Desktop" }
                { @(8..16) -contains $PSItem } { $Type = "Laptop" }
                Default { "Unknown" }
            }

            if ($AV.Count -gt 1) {
                $AntivirusProduct = $AV | Where-Object -FilterScript {$PSItem -ne 'Windows Defender'}
            }
            else {
                $AntivirusProduct = $AV
            }

            # Calculate age and reimage date
            $Age = (New-TimeSpan -Start ($Age.ConvertToDateTime($Age.ReleaseDate).ToShortDateString()) -End $(Get-Date)).Days / 365
            $ReimageDate = ($ReimageDate.ConvertToDateTime($ReimageDate.InstallDate).ToString("MM/dd/yyyy"))

            <# Calculate Decimal Age to Date #>
            $CalculatedAge = ($Age * 365.25)
            $CalculatedAge = (Get-Date).AddDays(-$CalculatedAge).ToString("MM/dd/yyyy")

            # Calculate Excel formula based off TODAYs date
            $ExcelAge = "=ROUND(YEARFRAC(`"$CalculatedAge`", TODAY()), 2)"

            # Big ol object dump
            $Properties = [PSCustomObject]@{
                AntiVirusProduct = $AntivirusProduct
                Hostname         = $Hostname
                Device           = $Make
                Type             = $Type
                Serial           = $Serial
                Edition          = $Edition
                OS               = $OS
                Memory           = $Memory
                Domain           = $Domain
                DecimalAge       = $Age
                Age              = $CalculatedAge
                Reimaged         = $ReimageDate
                ExcelAge         = $ExcelAge
            }

            # Add MAC if exists to object
            if ($EthMAC) { $Properties | Add-Member -NotePropertyName EthMAC -NotePropertyValue $EthMAC }
            if ($WlpMAC) { $Properties | Add-Member -NotePropertyName WlpMAC -NotePropertyValue $WlpMAC }

            # Append custom non-discoverable settings to end
            $CustomSettings = @{
                'Asset Tag'      = ""
                'Bios Password'  = ""
                'Boot Sequence'  = ""
            }
            $Properties | Add-Member -NotePropertyName Custom -NotePropertyValue $CustomSettings

            # Store computer info in DeviceArray, append additional devices
            [Void]$DeviceArray.Add($Properties)

            $Hosts   = (Get-ChildItem -Path "$FilePath").FullName
            Write-Verbose -Message "Variable [Hosts] set to $Hosts"
            if ($Hosts.Count -gt 0) {
                $HostsCheck = Get-Content -Path $Hosts -Raw | ConvertFrom-Json
                Write-Verbose -Message "Variable [HostsCheck] set to $HostsCheck"

                foreach ($Item in $HostsCheck) {
                    if ($Properties.Serial -eq $Item.Serial) {
                        Write-Verbose -Message "Removing duplicate entry"
                        Remove-Item -Path "$FilePath\$($Item.Hostname).json"
                    }
                }
            }
            
            $Properties | ConvertTo-Json | Out-File "$FilePath\$Computer.json"

            <# Prep specific properties for Excel pasting; tab delimited #>
            ($Properties | Select-Object * | ForEach-Object {
                $_.Device
                $_.Type
                $_.Hostname
                $_.Serial
                $_.Edition
                $_.OS
                $_.ExcelAge
                $_.Reimaged
            }) -join "`t" | Set-Clipboard -Append
        }
        catch [System.Exception] {
            $ExceptionLineNumber  = $PSItem.InvocationInfo.ScriptLineNumber
            $ExceptionLineContent = (Get-Content (Split-Path $MyInvocation.ScriptName -Leaf) -TotalCount $ExceptionLineNumber)[-1]
            $ExceptionMessage     = $PSItem.Exception.Message

            Write-Warning -Message "[ERROR] : Device [$Computer]
            $ExceptionMessage
            Exception caught on line $ExceptionLineNumber
            $ExceptionLineContent"
        }
    }
    
    <# Spicy STDOUT #>
    $DeviceArray | Format-Table -AutoSize
}

# If help, help please
if ($Help) { Get-Help; exit }

<# If there is no Hosts directory to store JSON, create it #>
$FilePath = $FilePath + "\Hosts"

if (!(Test-Path -Path "$FilePath")) {
    New-Item -Path "$FilePath" -Name "Hosts" -ItemType Container > $null
}


do {
    Show-Menu
    $Choice = Read-Host "Make a selection"
    Write-Verbose -Message "Variable [FilePath] set to $FilePath"

    switch ($Choice) {
        <# Prompt for AD search terms - only starting characters #>
        'a' { 
            $ADFilter = Read-Host "What computer would you like to search for?"
            $ADQuery = (Get-ADComputer -Filter "Name -like '$ADFilter*'" | Select-Object -ExpandProperty Name) -join ","
            $ADQuery = $ADQuery.Split(",").Trim(" ")
            Get-DeviceInfo -ComputerName $ADQuery
        }
        <# [l|L] Prompt for a computer to scan; exit on SIGINT #>
        <# [s|S] Prompt for a computer to scan; exit once complete #>
        { @('l','s') -contains $PSItem } {
            $HostnameExists = $false

            while (!$HostnameExists) {
                Write-Host "`nType 'stop|Stop|Ctrl+c' to exit.`n" -ForegroundColor Yellow

                # Get device from user; quit on stop
                $Computers = Read-Host "What computer would you like to scan?"
                if ($Computers -contains 'stop') { exit }

                # Combine all computers together
                $Computers = $Computers.Split(",").Trim(" ")
                Get-DeviceInfo -ComputerName $Computers

                # If user selected single scan, auto set to $true
                if ($PSItem -eq 's') { $HostnameExists=$true }
            }
        }
        'q' { Clear-Host; exit }
        default { Write-Warning -Message "Invalid menu choice" }  
    }
} while ($Choice -eq 'q')